pkg/services/ngalert/schedule/schedule.go

```diff
@@ -518,15 +531,21 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 						attribute.String("rule_fingerprint", fpStr),
 						attribute.String("tick", utcTick),
 					))
-					defer span.End()
 
-					evaluate(tracingCtx, f, attempt, ctx, span)
-					return nil
-				})
-				if err != nil {
-					logger.Error("Evaluation failed after all retries", "error", err)
+					noRetry := attempt == sch.maxAttempts
+					err := evaluate(tracingCtx, f, attempt, ctx, span, noRetry)
+					// This is extremely confusing - when we exhaust all retry attempts, or we have no retryable errors
+					// we return nil - so technically, this is meaningless to know whether the evaluation has errors or not.
+					span.End()
+					if err == nil {
+						return
+					}
+
+					logger.Error("Failed to evaluate rule", "version", ctx.rule.Version, "fingerprint", f, "attempt", attempt, "now", ctx.scheduledAt)
+					time.Sleep(retryDelay)
```

Comment:
I settle on the most simplistic approach because of the following:

The `New` method of the scheduler doesn't return an error, I couldn't support adding validation on the constructor that `max_attempts * retryDelay <= minIntervalDuration`. So, I tried adding it at runtime - the problem is that unless we factor in query duration there's still a chance that it can go `minIntervalDuration`.


On top of that,  Grafana never drops tricks - that means that no matter how slow an evaluation is, we'll make sure the next one (after the slow one is executed). Every delay approach has the potential to make sure we never catch up on evaluations under certain circumstances, but I guess this is the best we can do for now.

My preferred approach would have been to use something like exponential backoff and then simply cancel the retries if we're reaching the next evaluation, but this is too big of a change to backport all the way to 9.4.

If you have other suggestions, let me know.

Reply:
It wouldn't necessarily fix it, but I wonder if taking a `time.Now()` before starting evaluation and cancelling retry if we've hit `minIntervalDuration` would be enough to mitigate the problem for now. We still run the risk of the final evaluation going long, but we'd avoid continuing to retry well past the interval.

Reply:
I don't think it's worth it - as you said, we'll run the risk of the final evaluation going long aways and at most what we're saving is 1 evaluation and at the expense of making this more complicated to understand (we'll always retry the number of max attempts vs we'll sometimes retry the number of max attempts).